### PR TITLE
ntrn-252: feat: simplify config by fetching it on app setup

### PR DESCRIPTION
**Pull request by NeverHappened**
implements [this task](https://p2pvalidator.atlassian.net/browse/NTRN-252)

### CODE REVIEW

File: internal/app/app.go [hunk 1]
```diff
@@ -178,3 +200,80 @@ func loadChains(cfg config.NeutronQueryRelayerConfig, logRegistry *nlogger.Regis
...
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch target chain status: %w", err)
+	}
+
+	neutronStatus, err := neutronClient.Status(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch neutron chain status: %w", err)
+	}
+
+	// NOTE: waiting for connection to be created
```
Comment (pr0n00gler) [inline internal/app/app.go:229]: Why do we need to wait for the connection to be created?
I mean, when we start an ICQ relayer instance we expect that the necessary connection is created already, don't we?

Comment (NeverHappened) [reply internal/app/app.go:229]: Normally yes, but in case of integration tests no

File: internal/app/app.go [hunk 2]
```diff
@@ -3,6 +3,15 @@ package app
...
 	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/avast/retry-go/v4"
+
+	"github.com/neutron-org/neutron-query-relayer/internal/subscriber/querier/client/query"
+
+	rpcclienthttp "github.com/tendermint/tendermint/rpc/client/http"
```
Comment (sotnikov-s) [inline internal/app/app.go:14]: please run `goimports -w -local github.com/neutron-org .`

File: internal/app/app.go [hunk 3]
```diff
@@ -38,6 +48,13 @@ const (
 	KVProcessorContext           = "kv_processor"
 )
 
+// retries configuration for fetching connection info
+var (
+	RtyAtt = retry.Attempts(uint(5))
+	RtyDel = retry.Delay(time.Millisecond * 10000)
```
Comment (sotnikov-s) [inline internal/app/app.go:54]: 1. see no reasons for these variables to be public
2. `time.Millisecond * 10000` can be replaced with `time.Second * 10`

File: internal/app/app.go [hunk 4]
```diff
@@ -178,3 +200,80 @@ func loadChains(cfg config.NeutronQueryRelayerConfig, logRegistry *nlogger.Regis
...
+type connectionParams struct {
+	neutronChainID  string
+	neutronClientID string
+
+	targetChainID      string
+	targetClientID     string
+	targetConnectionID string
+}
+
+func loadConnParams(ctx context.Context, neutronClient, targetClient *rpcclienthttp.HTTP, restAddress string, neutronConnectionId string, logger *zap.Logger) (*connectionParams, error) {
```
Comment (sotnikov-s) [inline internal/app/app.go:213]: we have here **neutron**Client and **target**Client, but after them a simple `restAddress` which is ambiguous for it's not said which chain is meant

File: internal/app/app.go (hunk 1 above)
Comment (sotnikov-s) [reply internal/app/app.go:229]: this can be treat as connection health check, i.e. make sure the chain is sound and the address is configured properly

Comment (sotnikov-s) [reply internal/app/app.go:229]: what I only don't like here is:
1. this `NOTE:` — cmon it's a comment and therefore it's already a note, why leave such an obvious explanation;
2. the `client` variable name is deceptive: it's not a client, but a REST call response.

File: internal/app/app.go [hunk 5]
```diff
@@ -178,3 +200,80 @@ func loadChains(cfg config.NeutronQueryRelayerConfig, logRegistry *nlogger.Regis
...
+	// NOTE: waiting for connection to be created
+	var client *query.IbcCoreConnectionV1ConnectionOK
+	if err := retry.Do(func() error {
+		var err error
+
+		client, err = restClient.Query.IbcCoreConnectionV1Connection(&query.IbcCoreConnectionV1ConnectionParams{
+			ConnectionID: neutronConnectionId,
+			Context:      ctx,
+		})
+
```
Comment (sotnikov-s) [inline internal/app/app.go:238]: so many empty lines here :) could you please reduce their number?

**Review by sotnikov-s — CHANGES_REQUESTED** [review]

**Review by foxpy — CHANGES_REQUESTED** [review]

File: go.mod [hunk 6]
```diff
@@ -8,12 +8,15 @@ require (
 	github.com/cosmos/cosmos-sdk/api v0.1.0
 	github.com/cosmos/ibc-go/v3 v3.0.0
 	github.com/cosmos/relayer/v2 v2.0.0-beta7
+	github.com/docker/docker v20.10.7+incompatible
```
[the remark this answers is not in the document]
Comment (NeverHappened) [reply go.mod:11]: oh, that's strange

File: .env.example [hunk 7]
```diff
@@ -2,29 +2,17 @@
```
Comment (oopcode) [inline .env.example:2]: i like this diff sooooo much

**Review by oopcode — APPROVED** [review]
tests pass, & code is better. lgtm

File: README.md [hunk 8]
```diff
@@ -84,33 +84,28 @@ Relayer:
...
-| `RELAYER_TARGET_CHAIN_CHAIN_ID `                 | `string`          | target chain id                                                                                                                                                            | required |
 | `RELAYER_TARGET_CHAIN_ACCOUNT_PREFIX `           | `string`          | target chain account prefix                                                                                                                                                | required |
 | `RELAYER_TARGET_CHAIN_VALIDATOR_ACCOUNT_PREFIX ` | `string`          | target chain validator account prefix                                                                                                                                      | required |
 | `RELAYER_TARGET_CHAIN_TIMEOUT `                  | `time`            | timeout of target chain provider                                                                                                                                           | required |
```
Comment (foxpy) [inline README.md:101]: You have removed this parameter from example config yet it is specified as required in README :thinking:

**Review by foxpy — CHANGES_REQUESTED** [review]

File: .env.example [hunk 9]
```diff
@@ -2,29 +2,17 @@
...
 
 RELAYER_TARGET_CHAIN_RPC_ADDR=tcp://host.docker.internal:26657
-RELAYER_TARGET_CHAIN_CHAIN_ID=test-2
 RELAYER_TARGET_CHAIN_ACCOUNT_PREFIX=neutron
 RELAYER_TARGET_CHAIN_VALIDATOR_ACCOUNT_PREFIX=neutrondeveloper
-RELAYER_TARGET_CHAIN_TIMEOUT=1000s
-RELAYER_TARGET_CHAIN_CONNECTION_ID=connection-0
-RELAYER_TARGET_CHAIN_CLIENT_ID=07-tendermint-0
-RELAYER_TARGET_CHAIN_DEBUG=true
-RELAYER_TARGET_CHAIN_OUTPUT_FORMAT=json
```
Comment (sotnikov-s) [inline .env.example:27]: I understand why you removed chain ID, client ID and connection ID, but what's the point about the other fields? I can see you introduced default values for these parameters, but why aren't they in the example config files anymore?

Comment (oopcode) [reply .env.example:27]: these values are:

1. not necessary in `.env.example.dev`
2. necessary in `.env.example`, please restore them

File: internal/app/app.go (hunk 2 above)
Comment (NeverHappened) [reply internal/app/app.go:14]: I've already run it and it changed nothing(

File: README.md (hunk 8 above)
Comment (NeverHappened) [reply README.md:101]: Hm, added it back to .env.example

File: internal/config/config.go [hunk 10]
```diff
@@ -28,31 +28,26 @@ const EnvPrefix string = "RELAYER"
 type NeutronChainConfig struct {
 	RPCAddr        string        `required:"true" split_words:"true"`
 	RESTAddr       string        `required:"true" split_words:"true"`
-	ChainID        string        `required:"true" split_words:"true"`
 	HomeDir        string        `required:"true" split_words:"true"`
 	SignKeyName    string        `required:"true" split_words:"true"`
-	Timeout        time.Duration `required:"true" split_words:"true"`
+	Timeout        time.Duration `required:"true" split_words:"true" default:"10s"`
```
Comment (foxpy) [inline internal/config/config.go:33]: :thinking: why is it `required:"true"` if there is a default specified?

**Review by foxpy — CHANGES_REQUESTED** [review]

File: internal/config/config.go (hunk 10 above)
Comment (NeverHappened) [reply internal/config/config.go:33]: I think it does not matter much? if value is not present it won't fail anyway.

Comment (foxpy) [reply internal/config/config.go:33]: That's really strange but ok.

File: README.md [hunk 11]
```diff
@@ -84,33 +84,28 @@ Relayer:
...
-| `RELAYER_NEUTRON_CHAIN_CHAIN_ID `                | `string`          | neutron chain id                                                                                                                                                           | required |
 | `RELAYER_NEUTRON_CHAIN_HOME_DIR   `              | `string`          | path to keys directory                                                                                                                                                     | required |
 | `RELAYER_NEUTRON_CHAIN_SIGN_KEY_NAME`            | `string`          | key name                                                                                                                                                                   | required |
 | `RELAYER_NEUTRON_CHAIN_TIMEOUT `                 | `time`            | timeout of neutron chain provider                                                                                                                                          | required |
```
Comment (foxpy) [inline README.md:89]: This config value is not required anymore (and some other variables as well), please update README accordingly.

**Review by foxpy — CHANGES_REQUESTED** [review]

File: README.md (hunk 11 above)
Comment (NeverHappened) [reply README.md]: I've changed these fields with default to `optional`

Comment (foxpy) [reply README.md]: Thank you!

**Review by foxpy — DISMISSED** [review]

**Review by pr0n00gler — DISMISSED** [review]

**Review by sotnikov-s — APPROVED** [review]

**Review by foxpy — APPROVED** [review]

**Review by sotnikov-s — APPROVED** [review]

**Pull request merged**